Modernize Build...

lishaduck opened the pull request:
I just wanted an esm dist to make it debuggable...

---

Summary:
- Fixed a typo...
- Distributing ESM (non-breaking, as there isn't an API. Well, there technically was, but [a quick search](https://github.com/search?q=import+typesync+language%3AJavaScript&type=code&l=JavaScript) showed nobody used it, and it didn't seem like #90 wanted an API?)
- Build with Rolldown (via tsdown), isolated declaration emit using oxc
- ESLint 9 + flat config
- Vitest

~~Might break node 16?~~ (Yeah, rolldown doesn't support 16)

---

Summary part 2:
- Tune Vitest settings
- Fix #131
- A few small refactors
- Use tinyglobby instead of globby.
- Update readme (@jeffijoe, may I suggest rerecording the gif sometime?)

---

Summary pt. 3
- ESM-only bumps
- Fix bug cwd bug
- More cleanup

jeffijoe commented on src/__tests__/cli-util.test.ts:
```diff
@@ -1,6 +1,7 @@
 import { parseArguments } from '../cli-util'
 
-test('parseArguments', () => {
+test('parseArguments', ({ expect }) => {
```
`expect` can be imported from `vitest` too, is there any difference in doing this vs using the import?

lishaduck replied:
It can run asynchronously, so Vitest can parallelize the tests.
If it's imported globally, Vitest won't know which test is calling expect.

jeffijoe reviewed (approved):
This is awesome!

lishaduck commented:
Sooo... I was stuck in a car for a 7 hours today and sorta was bored...
Time for a push! I kinda fixed #131 and refactored the rest of everything in the process.

lishaduck commented:
Here ya go. Hopefully the rebase didn't break anything.

lishaduck commented:
Sorry for making you have to review this...

jeffijoe commented on src/fs-utils.ts:
```diff
@@ -1,26 +1,16 @@
-    if ((err as { code: string }).code === 'ENOENT') {
-      return false
+    /* v8 ignore next 3 */
+    if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
```
Nice

jeffijoe commented:
> Sorry for making you have to review this...

This is nothing compared to the amount of code I have to review at work (and I'm usually the worst offender of large PRs 😂 )

lishaduck commented:
Give me a minute, making more changes because of course 🤣

lishaduck commented:
Ok, added turborepo.*

* Disclaimer: I removed nodemon, but turbo's watching functionality is currently broken in single-package repos. I figure it'll probably get fixed someday.

jeffijoe commented:
Never used it before, what benefits do we gain from introducing it in a single-package repo? If it's nothing significant then I'd rather avoid the additional complexity.

lishaduck commented:
> Never used it before, what benefits do we gain from introducing it in a single-package repo? If it's nothing significant then I'd rather avoid the additional complexity.

I primarily use it to parallelize tasks. It also caches tasks (to speed up ci), and it (sometimes) has a watcher.

lishaduck commented:
Oh, and I found some esm todos I'm fixing. Again, give me a minute.

jeffijoe commented:
Let's revert the `turbo` change, I don't think the size of this repo justifies it. 😅

lishaduck commented:
You know, I should test this... instead of releasing this, could you prerelease this so I can double check it works on Windows? It did earlier, but then I switched globbing impls and probably ought to check it still works.

jeffijoe commented:
You should be able to `npm pack` it and install it locally, no?

lishaduck commented:
Yeah, but I don't have the typesync installed on my windows VM, just sheriff for the repro.

jeffijoe commented:
`typesync@0.13.5-0`

lishaduck commented:
Ok, glad I tested. Something broke (yes, it's that slow to test)

lishaduck commented:
🤦‍♂️ [`onlyfiles`](https://github.com/mrmlnc/fast-glob#onlyfiles) is true by default.

lishaduck commented:
Ok, this PR is good, but I found another bug, I'll probably push the fix if I find the issue.

lishaduck reviewed (commented):
Fixed the bug, CI's green.

lishaduck commented on src/cli.ts:
```diff
@@ -14,11 +14,12 @@ import type {
 } from './types'
 import { createWorkspaceResolverService } from './workspace-resolver'
 import * as fsUtils from './fs-utils'
+import packageJson from '../package.json' with { type: 'json' }
```
It's compatible with 18.20 (though rolldown bundles it).

lishaduck commented:
I was thinking about this and realized I'd thought of another bug in paths, which fixing should give a nice perf boost, so I'm going to do it.

lishaduck commented on package.json:
```diff
@@ -1,41 +1,39 @@
     "typesync": "./bin/typesync"
   },
-  "main": "lib/index.js",
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "default": "./dist/index.js"
+    }
+  },
```
Just double-checking—we actually want to expose this, right?

jeffijoe replied:
Yes

jeffijoe commented:
Lmk when you're done and I'll merge 👍

jeffijoe commented:
Released as 0.14, thanks @lishaduck !
